cassdegrees/ui/forms.py

```diff
@@ -0,0 +1,110 @@
+import json
+
+from api.models import DegreeModel, SubplanModel
+from django import forms
+from django.core.exceptions import NON_FIELD_ERRORS
+from django.forms import ModelForm
+
+
+class JSONField(forms.CharField):
+
+    def __init__(self, *args, field_id, **kwargs):
+        super().__init__(widget=forms.HiddenInput(attrs={'id': field_id}), *args, **kwargs)
+
+    def to_python(self, value):
+        return json.loads(value)
+
+
+class EditProgramFormSnippet(ModelForm):
+    # Use custom handlers for JSON fields
+    globalRequirements = JSONField(field_id='globalRequirements', required=False)
+
+    class Meta:
+        model = DegreeModel
+        fields = ('code', 'year', 'name', 'units', 'degreeType', 'globalRequirements')
+        widgets = {
+            'code': forms.TextInput(attrs={'class': "text tfull"}),
+            'year': forms.NumberInput(attrs={'class': "text tfull", 'type': "number"}),
+            'name': forms.TextInput(attrs={'class': "text tfull"}),
+            'units': forms.NumberInput(attrs={'class': "text tfull", 'type': "number"}),
+            # DegreeType auto generated
+        }
+        labels = {
+            'degreeType': "Program Type",
+        }
+        error_messages = {
+            NON_FIELD_ERRORS: {
+                'unique_together': "A program with the same %(field_labels)s already exists!",
+            }
+        }
+
+    def clean_code(self):
+        data = self.cleaned_data['code']
+        if len(data) < 4:
+            raise forms.ValidationError("This should be at least 4 characters!")
+        return data
+
+    def clean_year(self):
+        data = self.cleaned_data['year']
+        if data < 2000 or data > 3000:
+            raise forms.ValidationError("This should be between 2000-3000!")
+        return data
+
+    def clean_name(self):
+        data = self.cleaned_data['name']
+        if len(data) < 5:
+            raise forms.ValidationError("This should be at least 5 characters!")
+        return data
+
+
+class EditSubplanFormSnippet(ModelForm):
+    # Automatically injected by default
+    units = forms.IntegerField(widget=forms.HiddenInput(), required=False)
+
+    class Meta:
+        model = SubplanModel
+        fields = ('code', 'year', 'name', 'units', 'planType')
+        widgets = {
+            'code': forms.TextInput(attrs={'class': "text tfull"}),
+            'year': forms.NumberInput(attrs={'class': "text tfull", 'type': "number"}),
```

Comment:
```suggestion
            'year': forms.NumberInput(attrs={'class': "text tfull", 'type': "number", 'max': 3000}),
```

Restrict max value in the input value itself (Again, 'min' doesn't work)

Reply:
I'm not sure about double validation here - we already validate this serverside (see the clean_* methods). Thoughts?